README.md
# Monitor #

[![Build Status](https://travis-ci.org/opencloset/monitor.svg?branch=v1.1.28)](https://travis-ci.org/opencloset/monitor)

Dashboard in waiting room

## Version ##

v1.1.28

## 요구사항 ##

    $ sudo apt-get install redis-server
    $ npm install --global yarn
    $ yarn install

    $ yarn build    # for build frontend distribution
    $ yarn watch    # for development

    $ cpanm --mirror http://www.cpan.org \
        --mirror http://cpan.theopencloset.net \
        --installdeps .

    $ sqlite3 db/monitor.db < db/init.sql

### Install TTS mp3 files ###

- `public/tts/index`: `1번 탈의실로 입장해주세요`
  `1.mp3` ~ `15.mp3`
- `public/tts/room`: `1번 탈의실에 의류가 준비되었습니다`
  `1.mp3` ~ `15.mp3`

## 실행 ##

    $ ln -s monitor.conf.sample monitor.conf
    $ MOJO_CONFIG=monitor.conf morbo -l 'http://*:5000' scripts/monitor    # http://localhost:5000

## 환경변수 ##

- `OPENCLOSET_DATABASE_DSN`
- `OPENCLOSET_DATABASE_NAME`
- `OPENCLOSET_DATABASE_OPTS`
- `OPENCLOSET_DATABASE_PASS`
- `OPENCLOSET_DATABASE_USER`
- `OPENCLOSET_MONITOR_EMAIL`
  staff 서비스에 인증하기 위한 계정 이메일
- `OPENCLOSET_MONITOR_PASSWORD`
  staff 서비스에 인증하기 위한 계정 비밀번호
- `OPENCLOSET_NAVER_TTS_CLIENT_ID`
  https://www.ncloud.com/ 에서 추가한 애플리케이션(CSS)의 key
- `OPENCLOSET_NAVER_TTS_CLIENT_SECRET`
  https://www.ncloud.com/ 에서 추가한 애플리케이션(CSS)의 secret
- `OPENCLOSET_REDIS_URL` `redis://localhost:6379`
- `OPENCLOSET_STAFF_URL`
- `OPENCLOSET_WHITELIST`
  접속가능한 IP 목록
- `PORT`
  Listening port

## FAQ ##

### WRONGTYPE Operation against a key holding the wrong kind of value 오류 발생

실행 시 다음과 같은 오류가 발생할 수 있습니다.

```
[HGET opencloset:storage expiration] WRONGTYPE Operation against a key holding the wrong kind of value
```

이 경우 예전 버전의 `opencloset:storage` 키 유형 때문에 발생합니다.
다음 명령을 이용해서 해당 키를 제거한 다음 구동하면 정상 동작합니다.

```
$ redis-cli
127.0.0.1:6379> hget opencloset:storage expiration
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6379> del opencloset:storage
(integer) 1
127.0.0.1:6379> hkeys opencloset:storage
(empty list or set)
$
```

### Build docker image ###

    $ docker build -t opencloset/monitor .
    $ docker build -f Dockerfile.minion -t opencloset/monitor/minion .


Changes
# Change Log

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]
## [v1.1.28] - Sun, 30 May 2021 11:19:36 +0900
### Fixed
- job queue backend storage 의 오류발생에 대비해 enqueue 할때에 예외처리 추가 (#171)

## [v1.1.27] - 2021-05-17 05:35:19 Asia/Seoul
### Changed
- 착용여부 목록에서 "입고감+기록요망"을 "보관요망"으로 변경 (#228)

## [v1.1.26] - Sun, 28 Feb 2021 08:48:48 +0900
### Added
- 의류준비 화면에 주문서 태그 노출 - 웨딩촬영 및 본식 (#219)

## [v1.1.25] - Sun, 11 Oct 2020 21:38:27 +0900
### Changed
- 의류준비 화면에 여성의 의류품목에 원피스 추가 (#214)

## [v1.1.24] - Sun, 12 Apr 2020 11:45:32 +0900
### Added
- 의류준비 화면에 사용자 치수에 따른 열린옷장 기성복 사이즈 표시 (#206)

## [v1.1.23] - Sun, 16 Feb 2020 10:34:44 +0900
### Changed
- 영문이름은 자연스럽게 TTS 가 읽을 수 있도록 함 (#195)

## [v1.1.22] - Sat, 11 May 2019 14:20:50 +0900
### Changed
- 최근 대여의류의 자켓에 팔길이 추가 (#189)
- "입고감"에 "온라인예정" 추가

## [v1.1.21] - Tue, 07 May 2019 21:30:16 +0900
### Changed
- 대기실 화면에서 택배반납과 관련된 내용 삭제 (#193)

## [v1.1.20] - Sun, 21 Apr 2019 12:08:19 +0900
### Fixed
- 의류선택화면에서 기증이 없는 의류에 대한 예외처리 (#191)

## [v1.1.19] - Sun, 30 Sep 2018 12:48:40 +0900
### Fixed
- 탈의실 이름영역의 style 조정 (#181)

## [v1.1.18] - Sun, 16 Sep 2018 14:18:57 +0900
### Changed
- frontend build system 을 bower+grunt 에서 webpack 으로 변경
- 탈의실안내 화면의 대여자이름과 탈의실 번호를 더 크게 함 (#181)
- 탈의실안내 화면에 TTS 기능 강화 (#181)

## [v1.1.17] - Sat, 01 Sep 2018 17:04:43 +0900

### Added
- /room 에 TTS 기능 추가 (#181)

## [v1.1.16] - Fri, 03 Aug 2018 07:30:51 +0900

### Fixed
- `GET /room` 에서 발생하는 SecurityError 를 해결

## [v1.1.15] - Tue, 24 Jul 2018 13:39:50 +0900

### Added
- 변경 가능한 상태를 응답하는 `api#status` route 추가 (#183)

### Changed
- 탈의에서 의류준비가 된 주문서를 dashboard#room 에 표시

## [v1.1.14] - Wed, 11 Jul 2018 12:46:15 +0900

### Added
- 탈의실용 대시보드 추가 dashboard#room (#181)

## [v1.1.13] - Fri, 18 May 2018 11:28:16 +0900

### Changed
- TTS API 변경에 따른 패치적용 (#177)

## [v1.1.12] - Sat, 05 May 2018 16:32:21 +0900

### Changed
- `./db` 을 volume 으로 지정

## [v1.1.11] - Fri, 04 May 2018 20:47:48 +0900

### Fixed
- Set default config file in docker image

## [v1.1.10] - Fri, 04 May 2018 15:16:27 +0900

### Added
- travis-ci 에서 docker image 를 build

## [v1.1.9] - Thu, 22 Mar 2018 00:08:41 +0900

### Changed
- 의류준비탭의 입고감 버튼을 '입고감|기록요망|입고감+기록요망' 으로 확장해서 사용 (#168)

## [v1.1.8] - Wed, 21 Mar 2018 08:35:14 +0900

### Changed
- font-awesome js svg 대신에 css web-fonts 를 사용 (#176)

## [v1.1.7] - Mon, 19 Mar 2018 11:14:02 +0900

### Fixed
- 의류수선탭내의 토글버튼의 오동작 개선 (#176)

## [v1.1.6] - Wed, 14 Mar 2018 12:21:13 +0900

### Changed
- Upgrading font-awesome from version 4 to version 5
- 의류준비탭에서 '입고감' 토글 가능하도록 함 (#168)

## [v1.1.5] - Thu, 08 Mar 2018 12:36:47 +0900

### Added
- 방문으로 변경될때에 디버그용 로그 추가
- 방문예약 상태에서만 방문으로 상태를 변경할 수 있도록 예외처리

### Changed
- 취업날개 쿠폰 이용자는 바로 의류준비 상태로 넘어가지 않도록 함 (#175)

## [v1.1.4] - Sat, 03 Mar 2018 12:46:47 +0900

### Changed
- 탈의실이 11개에서 15개로 변경 (#172)

## [v1.1.3] - Thu, 23 Nov 2017 12:29:26 +0900

### Fixed
- typo

## [v1.1.2] - Tue, 21 Nov 2017 14:59:08 +0900

### Changed
- Display avg size instead of clothes-prediction (#165)
- 추천의류를 구하기 위해 평균값을 허리둘레를 사용 (#165)

## [v1.1.1] - Fri, 03 Nov 2017 02:10:33 +0900

### Fixed
- Add exception template for 500 error

## [v1.1.0] - Thu, 02 Nov 2017 16:10:27 +0900

### Added
- dockerize

### Changed
- job queue 의 로그메세지를 명확하게 함 (#161)
- haml 대신 ep 를 사용함

## [v1.0.5] - Wed, 16 Aug 2017 12:40:41 +0900

### Changed
- 구두 또는 타이만 대여할 경우 바로 의류준비 상태로 변경 (#160)

## [v1.0.4] - Fri, 11 Aug 2017 16:48:18 +0900

### Added
- 의류준비탭에 희망하는 지난대여이력을 표시 (#159)

## [v1.0.3] - Thu, 10 Aug 2017 17:02:01 +0900

### Added
- 대리인 대여의 경우 의류준비탭에 표시

## [v1.0.2] - Thu, 01 Jun 2017 15:31:38 +0900
### Changed
- 예약확인 탭의 자동완성 기능 변경

## [v1.0.1] - Thu, 11 May 2017 14:50:13 +0900
### Changed
- 의류준비탭의 탈의실 위치 변경 (#156)

v1.0.0   Fri, 05 May 2017 16:53:39 +0900
  - Naver TTS 를 사용해서 탈의실 입장을 안내함 (#154)

v0.9.9    Mon, 01 May 2017 17:54:42 +0900
  - 예약확인 탭의 안내글 수정 (#148)
  - 예약확인 탭의 스타일 변경 (#148)
  - 예약확인 탭에 숫자 키패드 추가 (#148)

v0.9.8    Wed, 19 Apr 2017 14:44:48 +0900
  - 매 실행마다 job queue reset
  - 의류추천 요청시간 10s -> 20s

v0.9.7    Wed, 19 Apr 2017 12:48:35 +0900
  - 의류추천의 중복된 enqueue 를 예방

v0.9.6    Wed, 19 Apr 2017 12:00:57 +0900
  - 의류추천 캐시의 expire 를 +1h 로 설정 (#153)

v0.9.5    Tue, 18 Apr 2017 19:03:32 +0900
  - 방문자 스스로 예약확인 할 수 있는 화면을 추가 (#148)
  - Queue 시스템 변경 (#151)
    Directory::Queue -> Minion

v0.9.4    Tue, 14 Mar 2017 12:58:40 +0900
  - jquery.timeago 형식을 UTC 에서 localtime 으로 변경

v0.9.3    Tue, 21 Feb 2017 18:06:07 +0900
  - 최근대여의류에 태그를 표시 (#144)

v0.9.2    Thu, 02 Feb 2017 11:09:56 +0900
  - 온라인 주문서에 booking date 가 없어서 발생하는 오류 해결 (#146)

v0.9.1    Mon, 09 Jan 2017 17:38:15 +0900
  - SMS 발신번호 변경
  - 추천의류에 태그를 표시 (#141)

v0.9.0    Fri, 25 Nov 2016 08:03:44 +0900
  - 온라인주문서 구분을 위해 online 컬럼을 사용

v0.8.11   Tue, 22 Nov 2016 17:22:32 +0900
  - 의류준비에 사용자정보 바로가기 링크 추가 (#129)

v0.8.10   Thu, 17 Nov 2016 15:10:55 +0900
  - 대시 보드 화면에서 500 오류 코드 발생 (#135)
  - /online 에 열린사진관 sms 매크로를 추가 (#137)
  - 여성의 목 둘레 대신에 허벅지 둘레를 표시 (#134)

v0.8.9    Mon, 29 Aug 2016 14:45:55 +0900
  - /target_date 를 인증없이 처리
  - 의류준비탭의 탈의실화면에 대여목적을 표시 (#128)

v0.8.8    Thu, 07 Jul 2016 18:57:51 +0900
  - 검색결과 정렬방식 개선 (#125)

v0.8.7    Thu, 07 Jul 2016 14:59:28 +0900
  - 의류준비 -> 의류준비중

v0.8.6    Mon, 04 Jul 2016 18:57:11 +0900
  - 빈 탈의실에 대해 상태 toggle (#112)

v0.8.5    Mon, 04 Jul 2016 14:20:09 +0900
  - 추천의류 스커트의 사이즈 표시변경 (#124)

v0.8.4    Fri, 24 Jun 2016 10:38:40 +0900
  - 의류검색 데몬의 표준출력 문제해결 (#94)

v0.8.3    Mon, 20 Jun 2016 18:31:31 +0900
  - 매번 의류검색하기전 인증절차 수행 (#118)
  - 온라인 SMS 매크로 내용 수정 (#123)
  - 비예약자의 예약시간을 잘보이게함 (#110)

v0.8.2    Thu, 16 Jun 2016 15:43:03 +0900
  - 지난의류의 스커트는 허리둘레도 표시

v0.8.1    Thu, 16 Jun 2016 14:07:05 +0900
  - 수선완료 목록의 삭제 상태변경 (#122)
  - 추천의류를 10개까지 보여줌

v0.8.0    Tue, 14 Jun 2016 18:33:15 +0900
  - preparation 리팩토링 (#94)
  - 제안의류(의류추천) 기능 추가 (#94)
  - redis 의 on/off 저장방식 대신에 직접 command(hset, hget..)를 사용

v0.7.0    Fri, 20 May 2016 18:53:16 +0900
  - rename OpenCloset::Status OpenCloset::Monitor::Status
  - 토글 가능한 대여의류품목 (#109)


lib/OpenCloset/Status.pm
package OpenCloset::Status;
use utf8;

our $STATUS_REPAIR        = 6;
our $STATUS_VISIT         = 13;
our $STATUS_MEASURE       = 16;
our $STATUS_SELECT        = 17;
our $STATUS_BOXING        = 18;
our $STATUS_PAYMENT       = 19;
our $STATUS_BOXED         = 44;
our $STATUS_DO_NOT_RENTAL = 40;

our $STATUS_FITTING_ROOM1  = 20;
our $STATUS_FITTING_ROOM2  = 21;
our $STATUS_FITTING_ROOM3  = 22;
our $STATUS_FITTING_ROOM4  = 23;
our $STATUS_FITTING_ROOM5  = 24;
our $STATUS_FITTING_ROOM6  = 25;
our $STATUS_FITTING_ROOM7  = 26;
our $STATUS_FITTING_ROOM8  = 27;
our $STATUS_FITTING_ROOM9  = 28;
our $STATUS_FITTING_ROOM10 = 29;
our $STATUS_FITTING_ROOM11 = 30;

our @ACTIVE_STATUS = (
    $STATUS_REPAIR,  $STATUS_VISIT,
    $STATUS_MEASURE, $STATUS_SELECT,
    $STATUS_BOXING,  $STATUS_PAYMENT,
    $STATUS_BOXED,   $STATUS_FITTING_ROOM1 .. $STATUS_FITTING_ROOM11
);

our %MAP = (
    $STATUS_REPAIR         => '수선',
    $STATUS_VISIT          => '방문',
    $STATUS_MEASURE        => '치수측정',
    $STATUS_SELECT         => '의류선택',
    $STATUS_BOXING         => '포장',
    $STATUS_BOXED          => '포장완료',
    $STATUS_PAYMENT        => '결제대기',
    $STATUS_DO_NOT_RENTAL  => '대여안함',
    $STATUS_FITTING_ROOM1  => '탈의01',
    $STATUS_FITTING_ROOM2  => '탈의02',
    $STATUS_FITTING_ROOM3  => '탈의03',
    $STATUS_FITTING_ROOM4  => '탈의04',
    $STATUS_FITTING_ROOM5  => '탈의05',
    $STATUS_FITTING_ROOM6  => '탈의06',
    $STATUS_FITTING_ROOM7  => '탈의07',
    $STATUS_FITTING_ROOM8  => '탈의08',
    $STATUS_FITTING_ROOM9  => '탈의09',
    $STATUS_FITTING_ROOM10 => '탈의10',
    $STATUS_FITTING_ROOM11 => '탈의11'
);

our %ORDER_MAP = (
    $STATUS_VISIT          => 1,
    $STATUS_MEASURE        => 2,
    $STATUS_SELECT         => 3,
    $STATUS_FITTING_ROOM1  => 4,
    $STATUS_FITTING_ROOM2  => 4,
    $STATUS_FITTING_ROOM3  => 4,
    $STATUS_FITTING_ROOM4  => 4,
    $STATUS_FITTING_ROOM5  => 4,
    $STATUS_FITTING_ROOM6  => 4,
    $STATUS_FITTING_ROOM7  => 4,
    $STATUS_FITTING_ROOM8  => 4,
    $STATUS_FITTING_ROOM9  => 4,
    $STATUS_FITTING_ROOM10 => 4,
    $STATUS_FITTING_ROOM11 => 4,
    $STATUS_REPAIR         => 5,
    $STATUS_BOXING         => 6,
    $STATUS_BOXED          => 6,
    $STATUS_PAYMENT        => 7,
);

our %REVERSE_ORDER_MAP = (
    '방문'       => 1,
    '치수측정' => 2,
    '의류선택' => 3,
    '탈의01'     => 4,
    '탈의02'     => 4,
    '탈의03'     => 4,
    '탈의04'     => 4,
    '탈의05'     => 4,
    '탈의06'     => 4,
    '탈의07'     => 4,
    '탈의08'     => 4,
    '탈의09'     => 4,
    '탈의10'     => 4,
    '탈의11'     => 4,
    '수선'       => 5,
    '포장'       => 6,
    '포장완료' => 6,
    '결제대기' => 7,
);

1;
